### PR TITLE
Add Context Parallel Support in Torchtune

### PR DESCRIPTION
**Pull request by SKRohit**
#### Context
The purpose of this PR is to
1. add context parallel support to lora and full finetuning recipe
2. update tests and/or documentation 

 

Our cp implementation is inspired from [torchtitan cp implementation](https://discuss.pytorch.org/t/distributed-w-torchtitan-breaking-barriers-training-long-context-llms-with-1m-sequence-length-in-pytorch-using-context-parallel/215082) and currently only supports non-packed datasets as sharding of mask (used for attention calculation in packed datasets) along sequence dimension needs some work. Also, our implementation does not shard rope cache along sequence dimension like in torchtitan instead it uses `input_pos`  to select right rope cache while applying rope embedding. Changes inside `padded_collate_sft` takes care of `input_pos` creation for non-packed datasets. Sequence length of batches change every iteration hence, we apply correct padding in each batch to ensure proper sharding of tokens across cp dims.

#### Changelog
* Add full finetuning with cp recipe: `recipes/configs/llama3_1/8B_full_cp.yaml`
* Add lora  finetuning with cp recipe: `recipes/configs/llama3_1/8B_lora_cp.yaml`
* Update `recipes/full_finetune_distributed.py` and `recipes/lora_finetune_distributed.py` to support cp 
* Update `padded_collate_sft` collate function to support data preparation for cp
* Update `torchtune/training/_distributed.py` to support cp mesh creation and `context_parallel` context creation
* Add unit test for padding with cp in `tests/torchtune/data/test_collate.py`
* Add unit test for context parallel tensor sharding in `tests/torchtune/training/test_distributed.py`
* * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * * *

### CODE REVIEW

File: recipes/configs/llama3_1/8B_lora.yaml [hunk 1]
```diff
@@ -35,6 +35,10 @@ model:
   lora_alpha: 16  # usually alpha=2*rank
   lora_dropout: 0.0
 
+# Parallelism
```
Comment (vivekgoe) [inline recipes/configs/llama3_1/8B_lora.yaml]: Can we create a new recipe instead of modifying existing one? Something like 8B_lora_cp.yaml

File: recipes/lora_finetune_distributed.py [hunk 2]
```diff
@@ -932,7 +992,7 @@ def _loss_step(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
 
         return loss
 
-    def validate(self) -> Dict[str, float]:
+    def validate(self) -> dict[str, float]:
```
Comment (vivekgoe) [inline recipes/lora_finetune_distributed.py]: why do we need this change?

File: recipes/lora_finetune_distributed.py [hunk 3]
```diff
@@ -968,7 +1028,7 @@ def validate(self) -> Dict[str, float]:
         log_dict = {"val_loss": avg_val_loss}
 
         if self._is_rank_zero:
-            log.info(f"Validation loss: {avg_val_loss:.4f}")
+            self._logger.info(f"Validation loss: {avg_val_loss:.4f}")
```
Comment (vivekgoe) [inline recipes/lora_finetune_distributed.py:1048]: Is this change needed due to a bug in base code?

File: recipes/lora_finetune_distributed.py [hunk 4]
```diff
@@ -142,9 +142,32 @@ def __init__(self, cfg: DictConfig) -> None:
...
+        data_shard = cfg.get("data_parallel_shard_dim", -1)  # -1 means to infer
+        data_replicate = cfg.get("data_parallel_replicate_dim", 1)
+        self.cp_degree = cfg.get("context_parallel_dim", 1)
+        self.context_parallel_rotate_method = cfg.get("context_parallel_rotate_method", "none")
+
+        # Set up n-d device mesh
+        self.parallel_dims = training.ParallelDims(
+            dp_replicate=data_replicate,
+            dp_shard=data_shard,
+            tp=1,
```
Comment (vivekgoe) [inline recipes/lora_finetune_distributed.py]: Why are we hard-coding tp? We should set it to self.tp_degree.  Since tp support is not integrated to this file, we can add an assert for now if tp_degree > 1.

File: recipes/lora_finetune_distributed.py [hunk 5]
```diff
@@ -915,9 +951,33 @@ def train(self) -> None:
     def _loss_step(self, batch: Dict[str, torch.Tensor]) -> torch.Tensor:
         # Shape [b, s], needed for the loss not the model
         labels = batch.pop("labels")
+        if self.cp_degree>1:
+            # For cp we need to pass the buffer
+            # and the buffer dim (cp_seq_dims) to apply cp
+            cp_buffers = [labels]
```
Comment (vivekgoe) [inline recipes/lora_finetune_distributed.py]: Why not start from an empty list?

File: recipes/lora_finetune_distributed.py [hunk 6]
```diff
@@ -915,9 +951,33 @@ def train(self) -> None:
...
+                if isinstance(v, torch.Tensor):
+                    cp_buffers.append(v)
+
+            # Create context parallel context if cp is enabled
+            optional_context_parallel_ctx = (
+            training.create_context_parallel_ctx(
+                cp_mesh=self.world_mesh["cp"],
+                cp_buffers=cp_buffers,
+                cp_seq_dims=[1]*len(cp_buffers),
+                cp_no_restore_buffers=set(cp_buffers),
```
Comment (vivekgoe) [inline recipes/lora_finetune_distributed.py]: Do we need a set?

File: recipes/lora_finetune_distributed.py [hunk 7]
```diff
@@ -915,9 +951,33 @@ def train(self) -> None:
...
+            cp_buffers = [labels]
+            for k,v in batch.items():
+                if isinstance(v, torch.Tensor):
+                    cp_buffers.append(v)
+
+            # Create context parallel context if cp is enabled
+            optional_context_parallel_ctx = (
+            training.create_context_parallel_ctx(
+                cp_mesh=self.world_mesh["cp"],
+                cp_buffers=cp_buffers,
```
Comment (vivekgoe) [inline recipes/lora_finetune_distributed.py]: TorchTitan seems to be adding more things to cp_buffers, can you please check what those are for? https://github.com/pytorch/torchtitan/blob/315191e95a5ca9f1005301a7e025cd6e97519ff0/torchtitan/train.py#L368

File: recipes/lora_finetune_distributed.py [hunk 8]
```diff
@@ -915,9 +951,33 @@ def train(self) -> None:
...
+                cp_no_restore_buffers=set(cp_buffers),
+                cp_rotate_method=self.context_parallel_rotate_method,
+            )
+            if self.parallel_dims.cp_enabled
+            else None
+            )
 
-        with self.activations_handling_ctx:
-            outputs = self._model(**batch)
+            with optional_context_parallel_ctx:
```
Comment (vivekgoe) [inline recipes/lora_finetune_distributed.py]: TorchTitan puts output computation, loss computation and loss.backward() under this "optional_context_parallel_ctx" context. Perhaps we should do the same, there maybe a reason for doing it that way.
https://github.com/pytorch/torchtitan/blob/315191e95a5ca9f1005301a7e025cd6e97519ff0/torchtitan/train.py#L401

It should be possible to move this context 1 level up in the call-stack.

**Review by vivekgoe — CHANGES_REQUESTED** [review]

File: recipes/lora_finetune_distributed.py (hunk 3 above)
Comment (SKRohit) [reply recipes/lora_finetune_distributed.py:1048]: yes

File: recipes/lora_finetune_distributed.py (hunk 2 above)
Comment (SKRohit) [reply recipes/lora_finetune_distributed.py]: current main torchtune branch has this. Later while pushing pr to `pytorch/torchtune` we may need to do this change.

File: recipes/lora_finetune_distributed.py (hunk 4 above)
Comment (SKRohit) [reply recipes/lora_finetune_distributed.py]: yes. I did it so that any one reading this script doesn't get confused that tp can be supported by setting tp_degree value in yaml file

File: recipes/lora_finetune_distributed.py (hunk 5 above)
Comment (SKRohit) [reply recipes/lora_finetune_distributed.py]: we can. Could be reformatted.

File: recipes/configs/llama3_1/8B_lora.yaml (hunk 1 above)
Comment (SKRohit) [reply recipes/configs/llama3_1/8B_lora.yaml]: yes

File: recipes/lora_finetune_distributed.py (hunk 6 above)
Comment (SKRohit) [reply recipes/lora_finetune_distributed.py]: yes as per `context_parallel` signature. https://github.com/pytorch/pytorch/blob/2585960b47d6429a119665216123cb00435efde9/torch/distributed/tensor/experimental/_attention.py#L1342C1-L1347C60
```
def context_parallel(
    mesh: DeviceMesh,
    *,
    buffers: Optional[list[torch.Tensor]] = None,
    buffer_seq_dims: Optional[list[int]] = None,
    no_restore_buffers: Optional[set[torch.Tensor]] = None,
```

File: recipes/lora_finetune_distributed.py (hunk 8 above)
Comment (SKRohit) [reply recipes/lora_finetune_distributed.py]: Outside cp context `buffers` are unsharded again so other operations on `buffers` needs to be moved inside context if those `buffers` are not passed to `cp_no_restore_buffers`.  tensors (or buffers) passed to `cp_no_restore_buffers` are not restored outsid cp context. Activations created during forward remain sharded as per cp mesh once cp context closed, so `loss.backward()` could be moved outside context but will verify. Also, verify why TorchTitan puts loss and backward inside cp context.

File: recipes/lora_finetune_distributed.py (hunk 7 above)
Comment (SKRohit) [reply recipes/lora_finetune_distributed.py]: Extra part that is being passed is rotation matrix (shape (seq_len, attention_head_dim/2) with positional frequencies i.e. 

image

It is used to calculate rotary embedding in polar form. I missed handling it in my changes. For `packed=False` case sharding `pos_embeddings.cache` on dim 0 should work as it similar to sharding `freqs_cis` in torchtitan code. For packed = True some post processesing of mask, and input ids will be required to handle cp.

File: recipes/lora_finetune_distributed.py (hunk 8 above)
Comment (bhargaveede) [reply recipes/lora_finetune_distributed.py]: Isn't is better to have those buffers sharded to save some memory. If we unshard, won't we keep same copy in cp_degree ranks?

Comment (SKRohit) [reply recipes/lora_finetune_distributed.py]: yes currently because I am passing those buffers to cp_no_restore_buffers they are not being restored after sharding.

File: recipes/lora_finetune_distributed.py (hunk 7 above)
Comment (bhargaveede) [reply recipes/lora_finetune_distributed.py]: As discussed, We need to slice the cache based on dim if `cache` is full_size.

File: recipes/lora_finetune_distributed.py (hunk 2 above)
Comment (vivekgoe) [reply recipes/lora_finetune_distributed.py]: Please remove it for now. When you rebase your PR later you will automatically get this.

File: recipes/lora_finetune_distributed.py (hunk 4 above)
Comment (vivekgoe) [reply recipes/lora_finetune_distributed.py]: can we read from self.tp_degree and assert if its >1 with a message saying tp not support for now.

File: recipes/configs/llama3_1/8B_lora.yaml (hunk 1 above)
Comment (vivekgoe) [reply recipes/configs/llama3_1/8B_lora.yaml]: Please change, it will be easier if we keep refining and making it closer to final PR we want.

File: recipes/lora_finetune_distributed.py (hunk 3 above)
Comment (vivekgoe) [reply recipes/lora_finetune_distributed.py:1048]: If it is fixed in latest torchtune, better to remove this and get this when we rebase.

File: torchtune/models/llama3_1/_model_builders.py [hunk 9]
```diff
@@ -90,6 +90,7 @@ def lora_llama3_1_8b(
     lora_dropout: float = 0.0,
     use_dora: bool = False,
     quantize_base: bool = False,
+    max_seq_len: int = 131072,
```
Comment (bhargaveede) [inline torchtune/models/llama3_1/_model_builders.py]: What is the value of max_seq_len if nothing is given in yaml?
Check it's 131072 and this extra signature change isn't creating problem.

Comment (SKRohit) [reply torchtune/models/llama3_1/_model_builders.py]: No because in the yaml I am passing the `max_seq_len` value under Model Arguments. Also, I have added an assert that `tokenizer.max_seq_len` must be equal to `model.max_seq_len`.

File: recipes/lora_finetune_distributed.py (hunk 7 above)
Comment (SKRohit) [reply recipes/lora_finetune_distributed.py]: handling it in new changes.

File: recipes/lora_finetune_distributed.py (hunk 8 above)
Comment (SKRohit) [reply recipes/lora_finetune_distributed.py]: moved loss calculation and loss.backward under cp context, so no unshard of buffers is happening.

File: recipes/lora_finetune_distributed.py (hunk 4 above)
Comment (SKRohit) [reply recipes/lora_finetune_distributed.py]: yes. changes done

File: recipes/lora_finetune_distributed.py [hunk 10]
```diff
@@ -142,9 +143,38 @@ def __init__(self, cfg: DictConfig) -> None:
...
-
         self._is_rank_zero = self.rank == 0
+        data_shard = cfg.get("data_parallel_shard_dim", -1)  # -1 means to infer
+        data_replicate = cfg.get("data_parallel_replicate_dim", 1)
+        self.cp_degree = cfg.get("context_parallel_dim", 1)
+        self.context_parallel_rotate_method = cfg.get("context_parallel_rotate_method", "none")
+        self.tp_degree = cfg.get("tensor_parallel_dim", 1)
+        assert (self.tp_degree == 1), "Tensor parallelism is not supported in this recipe. Please set tensor_parallel_dim to 1."
+        if self.cp_degree > 1:
+            assert not (cfg.model.get("max_seq_len")) and (cfg.tokenizer.max_seq_len == cfg.model.max_seq_len), (
```
Comment (bhargaveede) [inline recipes/lora_finetune_distributed.py]: not shouldn't be there

Comment (SKRohit) [reply recipes/lora_finetune_distributed.py]: fixed it

**Comment by SKRohit** [comment]
These changes have been added in torchtune repo: https://github.com/pytorch/torchtune/pull/2879
